content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -889,6 +853,177 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
+[source,groovy]
+----
+def username = 'Jenkins'
+echo 'Hello Mr. ${username}'
+echo "I said, Hello Mr. ${username}"
+----
+
+Would result in:
+
+[source]
+----
+Hello Mr. ${username}
+I said, Hello Mr. Jenkins
+----
+
+Understanding how to use string interpolation is vital for using some of
+Pipeline's more advanced features.
+
+==== Interpolation of sensitive environment variables
+
+[WARNING]
+======
+Groovy string interpolation should [red]*never* be used with credentials.
+======
+
+Groovy string interpolation can leak sensitive environment variables (i.e. credentials, see: <<Handling credentials>>).
+This is because the sensitive environment variable will be interpolated during Groovy evaluation, and the environment variable's value could be made available earlier than intended, resulting in sensitive data leaking in various contexts.
+
+For example, consider a sensitive environment variable passed to the `sh` step.
+
+```
+pipeline {
+    agent any
+    environment {
+        EXAMPLE_CREDS = credentials('example-credentials-id')
+    }
+    stages {
+        stage('Example') {
+            steps {
+                /* WRONG! */
+                sh("curl -u ${EXAMPLE_CREDS_USR}:${EXAMPLE_CREDS_PSW} https://example.com/")
+            }
+        }
+    }
+}
+```
+Should Groovy perform the interpolation, the sensitive value will be injected directly into the arguments of the `sh` step, which among other issues, means that the literal value will be visible as an argument to the `sh` process on the agent in OS process listings.
+Using single-quotes instead of double-quotes when referencing these sensitive environment variables prevents this type of leaking.
+
+```
+pipeline {
+    agent any
+    environment {
+        EXAMPLE_CREDS = credentials('example-credentials-id')
+    }
+    stages {
+        stage('Example') {
+            steps {
+                /* CORRECT */
+                sh('curl -u $EXAMPLE_CREDS_USR:$EXAMPLE_CREDS_PSW https://example.com/')
+            }
+        }
+    }
+}
+```
+
+
+==== Injection via interpolation
+
+[WARNING]
+======
+Groovy string interpolation can inject rogue commands into command interpreters via special characters.
+======
+
+Another note of caution. Using Groovy string interpolation for user-controlled variables with steps that pass their arguments to command interpreters such as the `sh`, `bat`, `powershell`, or `pwsh` steps can result in problems analogous to SQL injection.
+This occurs when a user-controller variable (generally an environment variable, usually a parameter passed to the build) that contains special characters (e.g. `' " / \ $ & % ^ > < | ;`) is passed to the `sh`, `bat`, `powershell`, or `pwsh` steps using Groovy interpolation.
+For a simple example:
+
+```
+pipeline {
+  agent any
+  parameters {
+    password(name: 'EXAMPLE_SECRET', defaultValue: 'sec>ret', description: 'Enter a password')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        sh "echo $EXAMPLE_SECRET"
```

Comment:
Discussed elsewhere, but for this example I would use a plain `string` parameter to make it clear that this problem isn't directly related to whether the value being interpolated is a secret or not.

Reply:
updated to string parameter in 3481e9984ce629aa54dab6a8d64dac087f3fc4b3

Reply:
Looks like this is still a password parameter? (same with the corrected example)